lib/rubocop/cop/lint/missing_super.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for the presence of constructors and lifecycle callbacks
+      # without calls to `super`.
+      #
+      # @example
+      #   # bad
+      #   class Employee < Person
+      #     def initialize(name, salary)
+      #       @salary = salary
+      #     end
+      #   end
+      #
+      #   # good
+      #   class Employee < Person
+      #     def initialize(name, salary)
+      #       super(name)
+      #       @salary = salary
+      #     end
+      #   end
+      #
+      #   # bad
+      #   class Parent
+      #     def self.inherited(base)
+      #       do_something
+      #     end
+      #   end
+      #
+      #   # good
+      #   class Parent
+      #     def self.inherited(base)
+      #       super
+      #       do_something
+      #     end
+      #   end
+      #
+      class MissingSuper < Base
+        CONSTRUCTOR_MSG = 'Call `super` to initialize state of the parent class.'
+        CALLBACK_MSG    = 'Call `super` to invoke callback defined in the parent class.'
+
+        STATELESS_CLASSES = %w[BasicObject Object].freeze
+
+        OBJECT_LIFECYCLE_CALLBACKS  = %i[method_missing respond_to_missing?].freeze
+        CLASS_LIFECYCLE_CALLBACKS   = %i[inherited].freeze
+        MODULE_LIFECYCLE_CALLBACKS  = %i[included extended prepended].freeze
+
+        def on_def(node)
+          return if should_not_check?(node)
+
+          if node.method?(:initialize)
+            add_offense(node, message: CONSTRUCTOR_MSG) if inside_class_with_stateful_parent?(node)
+          elsif callback_method_def?(node)
+            add_offense(node, message: CALLBACK_MSG)
+          end
+        end
+
+        def on_defs(node)
+          return if !callback_method_def?(node) || contains_super?(node)
+
+          add_offense(node, message: CALLBACK_MSG)
+        end
+
+        private
+
+        def should_not_check?(node)
+          !(node.method?(:initialize) || callback_method_def?(node)) ||
+            contains_super?(node)
+        end
```

Comment:
I was a bit surprised that this predicate is inverted. It also means it can't short circuit, but must evaluate everything. Also the name makes it sound like it's a pre-check, but it's also the actual check itself. Compare, e.g.:

```ruby
def offender?(node)
  (node.method?(:initialize) || callback_method_def?(node)) && !contains_super?(node)
end
```

Reply:
Yes, thanks, much better.